src/OrleansRuntime/LogViews/NotificationTracker.cs

```diff
@@ -0,0 +1,203 @@
+using Orleans.MultiCluster;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Orleans.Runtime.LogViews
+{
+
+    /// <summary>
+    /// Helper class for tracking notifications that a log view grain sends to other clusters after updating the log.
+    /// </summary>
+    internal class NotificationTracker
+    {
+        private IProtocolServices services;
+        private Dictionary<string, NotificationWorker> sendworkers;
+        private int maxnotificationbatchsize;
+
+        public NotificationTracker(IProtocolServices services, MultiClusterConfiguration configuration, int maxnotificationbatchsize)
+        {
+            this.services = services;
+            sendworkers = new Dictionary<string, NotificationWorker>();
+            this.maxnotificationbatchsize = maxnotificationbatchsize;
+
+            foreach (var x in configuration.Clusters)
+                if (x != services.MyClusterId)
+                {
+                    services.Verbose("Now sending notifications to {0}", x);
+                    sendworkers.Add(x, new NotificationWorker(services, x, maxnotificationbatchsize));
+                }
+        }
+
+        public void BroadcastNotification(INotificationMessage msg, string exclude = null)
+        {
+            foreach (var kvp in sendworkers)
+                if (kvp.Key != exclude)
+                {
+                    var w = kvp.Value;
+                    w.Enqueue(msg);
+                }
+        }
+
+        /// <summary>
+        /// last observed exception, or null if last notification attempts were successful for all clusters
+        /// </summary>
+        public Exception LastException
+        {
+            get
+            {
+                return sendworkers.Values.OrderBy(ns => ns.LastFailure).Select(ns => ns.LastException).LastOrDefault();
+            }
+        }
+
+        /// <summary>
+        /// Update the multicluster configuration (change who to send notifications to)
+        /// </summary>
+        public void ProcessConfigurationChange(MultiClusterConfiguration oldconf, MultiClusterConfiguration newconf)
+        {
+            var removed = sendworkers.Keys.Except(newconf.Clusters);
+            foreach (var x in removed)
+            {
+                services.Verbose("No longer sending notifications to {0}", x);
+                sendworkers[x].Done = true;
+                sendworkers.Remove(x);
+            }
+
+            var added = oldconf == null ? newconf.Clusters : newconf.Clusters.Except(oldconf.Clusters);
+            foreach (var x in added)
+                if (x != services.MyClusterId)
+                {
+                    services.Verbose("Now sending notifications to {0}", x);
+                    sendworkers.Add(x, new NotificationWorker(services, x, maxnotificationbatchsize));
+                }
+        }
+
+
+        public enum NotificationQueueState : byte
```

Comment:
Naively I am assuming byte is better because it uses less storage. If anyone has any evidence to the contrary, let me know.

I did not actually know about using byte for enums until @jdom asked me to add it to one of the enums in one of my other PRs.